src/utils/types.ts

```diff
@@ -22,3 +22,33 @@ export type RequiredId<T extends { [key in "id"]?: string }> = Replace<T, "id",
 export type Replace<T, Key extends keyof T, NewValue, IsOptional extends boolean = false> =
   & Omit<T, Key>
   & (IsOptional extends true ? { readonly [key in Key]?: NewValue } : { readonly [key in Key]: NewValue });
+
+/**
+ * Extracts a transformed `ReadonlyMap` where the values are simplified to the `selfId` property
+ * of the objects if available. Otherwise returns the type as is.
+ *
+ * @template T - The type to transform, e.g. `ReadonlyMap<string, object>`.
+ * @returns A `ReadonlyMap` keyed similarly but with values as the `selfId` of the objects,
+ *          or the original type `T` if `T` is not a `ReadonlyMap` or the objects don't have `selfId`.
+ */
+export type ExtractSelfId<T> = T extends ReadonlyMap<string, infer U>
+  ? U extends { selfId: infer V } ? ReadonlyMap<string, V>
+  : T
+  : T;
+
+export type Transformable<T> = T extends object ?
```

Comment:
I am not exactly sure why we need this type

Reply:
It's just a recursive definition of a transformable type with index signature, to avoid casting to `any` in the reduce.

Reply:
Okay, now I see what you wanted to achieve. You wanted to preserve the type when transforming the object. 

This type is used only in `traverseAndReplace`, which is used in `transformReferences` in file `reference.ts`

However, from the type point of view, it's not that easy.  Moreover, from the type point of view, your solution is not kind of correct, therefore you needed to use `any`

`traverseAndReplace` always expects objects and in some cases, you send it to the function which is in runtime string but you retyped it into transformable<string>. Then you are doing type checking whether it is a string, which again in runtime is correct but it is a string, however, the type system is kind of confused (see never type in the below picture)

<img width="528" alt="image" src="https://github.com/kontent-ai/data-ops/assets/54802833/c24f67f9-7efc-408e-b1b7-405d5511c457">

The same is for Array, you said in the function definition that the value would be object, and it is Array (even though I am not sure what happens there as Array is also an object, but I believe it is something workaroundish also)

From the bigger picture point of view, transformReferences takes the object and needs to return the object of same type (it only replaces ids to other ids)

Even though it is nice you tried to make a general solution, in this case I think it would be good enough to keep simpler solution you can make something like this: 

```ts
export const transformReferences = <T extends object, U extends keyof ImportContext>(
  object: T,
  flatContext: SimplifyContext | SimplifyContext<U>,
): T => {
  /**
   * this merges all simplified context maps into a single map, which is subsequently used
   * for lookup of transformed values. if no match is found, original value is returned.
   *
   * if the value is an `object`, transforms it recursively.
   */
  const flatContextMap = new Map<string, string>(Object.values(flatContext).flatMap(c => [...c]));

  const traverseAndReplace = (value: unknown): unknown => {
    if(typeof value === 'string'){
      return flatContextMap.get(value) ?? value
    }

    if(Array.isArray(value)){
      return value.map((item) => traverseAndReplace(item));
    }

    if(value !== null  && typeof value === 'object') {
      return Object.fromEntries(Object.entries(value).map(([key, value]) => [key, traverseAndReplace(value)]));
    }

    return value;
  };

  return traverseAndReplace(object) as T;
};
```
In this state, it is still not very usable everywhere in general, as it might return oldId when it is not found in the context. That could result in more difficulties when debugging - we might see there is an error and then expect that is due to the oldId (that does not exist in the target environment) while the right error would be totally different. However, this is only assumption what might happen, rather what is the actual behavior.

On the other hand, I like that if some references are added (for instance references to collections in webhooks) this solution would handle it, even without us knowing :)

I also wonder if we can use this function in other places than just webhooks (such as content type, options in multiple-choice elements, or even referencing snippet in snippet element), whether it makes sense to have it exported it or have it just inside the webhook entity.

What do you think?